kvserver: avoid exceeding `TargetBytes`

erikgrinaker opened the pull request:
**kvserver: avoid exceeding `TargetBytes`**

Previously, `Header.TargetBytes` would always overshoot the target, i.e.
a scan would terminate only after the result set hit or exceeded
`TargetBytes`.

This patch changes the behavior to never overshoot, except for when the
first result exceeds `TargetBytes` (to make sure clients can always make
progress). However, this is not backwards-compatible with the 21.2
`DistSender` logic, which relies on responses overshooting `TargetBytes`
when enforcing limits for split batches. It therefore adds a
corresponding `TargetBytesAvoidExcess` version gate that only enables
this for 22.1, and `MVCCScanOptions.TargetBytesAvoidExcess` to vary the
scan behavior based on it.

Release note: None

**kvserver: add `TargetBytesAllowEmpty` option**

The `Header.TargetBytes` request limit will be overshot if the first
result exceeds it, to ensure the client can make progress. However, an
upcoming [parallel get/scan streaming library][1] needs to be able to
disable this behavior to avoid exceeding memory budgets when dispatching
parallel requests.

This patch adds a field `TargetBytesAllowEmpty` which will allow an
empty response if the first result exceeds `TargetBytes`. This only has
an effect for `Get` and `Scan` requests, and only on 22.1 clusters since
21.2 `DistSender` logic relies on the limit always being overshot. It
is not supported for `Export` requests where it will return an error.

[1]: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md

Resolves #68050.

Release note: None

**kvserver: add `ResponseHeader.ResumeNextBytes`**

This patch adds the field `ResponseHeader.ResumeNextBytes` which may be
populated with the size of the next result in the `ResumeSpan` if
`ResumeReason` is `RESUME_BYTE_LIMIT` (i.e. if `TargetBytes` was
exceeded). This will be needed by an upcoming [parallel get/scan
streaming library][1] to better manage its memory budget.

The field is best-effort, and may be omitted in some cases where the
result size exactly equals `TargetBytes` and we would have to do
additional work to obtain `ResumeNextBytes` (e.g. at the end of a range,
where we would have to send an additional RPC to the next range).

[1]: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md

Release note: None

**kvclient: rename replyResults to replyKeys**

Release note: None

erikgrinaker committed:
kvserver: add `TargetBytesAllowEmpty` option

The `Header.TargetBytes` request limit will be overshot if the first
result exceeds it, to ensure the client can make progress. However, an
upcoming [parallel get/scan streaming library][1] needs to be able to
disable this behavior to avoid exceeding memory budgets when dispatching
parallel requests.

This patch adds a field `TargetBytesAllowEmpty` which will allow an
empty response if the first result exceeds `TargetBytes`. This only has
an effect for `Get` and `Scan` requests, and only on 22.1 clusters since
21.2 `DistSender` logic relies on the limit always being overshot. It
is not supported for `Export` requests where it will return an error.

[1]: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68370)
<!-- Reviewable:end -->

pbardea commented:
Bulk/export request changes LGTM -- removing myself from reviewers since the change is small.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, and @nvanbenschoten)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBclJo8WNIlBiVE_oU:-MgBclJo8WNIlBiVE_oV:b6hc0r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/kv/kvserver/batcheval/cmd_get.go#L61)):*
> ```Go
> 	}
> 	if val != nil {
> 		fmt.Printf("size: %v\n", len(val.RawBytes))
> ```

nit: seems like a leftover.
___
*[pkg/roachpb/api.proto, line 2029 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBd8ln8K0cil7oCztv:-MgBd8loB7WfMpmpbX4s:b-panmsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/roachpb/api.proto#L2029)):*
> ```protobuf
>   // particular, it will return an empty response if the first result exceeds
>   // the target. Not supported for ExportRequest.
>   bool strict_target_bytes = 21;
> ```

It would also be great to somehow propagate the information about the size of the KV that was fetched but exceeded the strict target bytes (so that the `Streamer` would know the minimum budget necessary to get at least one KV for the span). 
___
*[pkg/storage/testdata/mvcc_histories/target_bytes, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBdqTu4TwG7EezPPKH:-MgBdqTu4TwG7EezPPKI:byjrhy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/storage/testdata/mvcc_histories/target_bytes#L117)):*
> ```
> run ok
> # Target size one byte returns one result (overshooting instead of returning
> # nothing), unless stricttargetsize is enabled. Upping the target size
> ```

nit: `s/stricttargetsize/stricttargetbytes/`.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBclJo8WNIlBiVE_oU:-MgBmMdH09nJF4IK8F_Y:bog5knh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/kv/kvserver/batcheval/cmd_get.go#L61)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: seems like a leftover.
</blockquote></details>

Oops, indeed.
___
*[pkg/roachpb/api.proto, line 2029 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBd8ln8K0cil7oCztv:-MgBmQ_v25Zq9xW4GzGr:b-ifjv3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/roachpb/api.proto#L2029)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It would also be great to somehow propagate the information about the size of the KV that was fetched but exceeded the strict target bytes (so that the `Streamer` would know the minimum budget necessary to get at least one KV for the span). 
</blockquote></details>

Right, that might actually come in handy to signal the DistSender that a limit was exceeded too (see comment in PR description). Will look into it.
___
*[pkg/storage/testdata/mvcc_histories/target_bytes, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBdqTu4TwG7EezPPKH:-MgBmeDXBQZEmTragliD:b-iry7x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/storage/testdata/mvcc_histories/target_bytes#L117)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/stricttargetsize/stricttargetbytes/`.
</blockquote></details>

Done, thanks


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1302 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgC5I1f326fgJBemXhb:-MgC5I1gAFMelW467FbL:b3q48e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1302)):*
> ```Go
> 					replyResults += r.GetInner().Header().NumKeys
> 					replyBytes += r.GetInner().Header().NumBytes
> 					if span := r.GetInner().Header().ResumeSpan; span != nil && span.Valid() {
> ```

Just a random thought I had: does it matter performance-wise that we're calling `r.GetInner().Header()` three times in a row? Would it be faster to store a local variable for it?


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I think this would also be a good time to think through whether we want `StrictTargetBytes = false` to allow overshoot even after it has gathered a single key-value. I say this because the decision about that may slightly alter the meaning of the new `StrictTargetBytes` flag. Right now, the meaning is consistent, though "strict" is a little vague. But if we make the standard `TargetBytes` a little stricter, then the flag added here may be more precisely framed as `TargetBytesAllowsEmpty` or something along those lines. Tightening that up would be a nice win. Of course, any changes to `TargetBytes` will be backwards incompatible (given the DistSender logic), so we'd need to make them carefully. 

> Reviewed 14 of 16 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MgCRICc4TAWQiytLMq6:b-slfhp5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, and @yuzefovich)

___
*[pkg/ccl/storageccl/export.go, line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCPGRAC7z7WsRO1jP9:-MgCPGRAC7z7WsRO1jPA:b-1axc7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/ccl/storageccl/export.go#L91)):*
> ```Go
> 
> 	if h.StrictTargetBytes {
> 		return result.Result{}, errors.Errorf("StrictTargetBytes not supported for ExportRequest")
> ```

`errors.AssertionFailedf`
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1302 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgC5I1f326fgJBemXhb:-MgCPKS22MysRVLg0ha8:b-ui8fsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1302)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Just a random thought I had: does it matter performance-wise that we're calling `r.GetInner().Header()` three times in a row? Would it be faster to store a local variable for it?
</blockquote></details>

Yeah, this is a good point. Let's pull out the header into a local variable to avoid the repeat type switch and dynamic dispatch.
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCPzrM9Y4pCZ9GHhdb:-MgCPzrM9Y4pCZ9GHhdc:bchsjui) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1315)):*
> ```Go
> 					ba.MaxSpanRequestKeys -= replyResults
> 					// Exiting; any missing responses will be filled in via defer().
> 					if ba.MaxSpanRequestKeys == 0 {
> ```

Can this condition and the one below both be replaced by `hasResumeSpan`?
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCQFiKBEXtkaP3eS5d:-MgCQFiKBEXtkaP3eS5e:b-a64zc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvserver/batcheval/cmd_get.go#L59)):*
> ```
> int64(len(val.RawBytes))
> ```

Could we pull this into a local variable and then leave a comment on it that this is different than the contribution of key-values towards Scan's budget because a Get does not return a key-values key?
___
*[pkg/kv/kvserver/batcheval/cmd_get_test.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCQwa508gGJotN3lzZ:-MgCQwa508gGJotN3lz_:bvjldqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvserver/batcheval/cmd_get_test.go#L59)):*
> ```Go
> 		{targetBytes: 1, expectResume: false},
> 		{targetBytes: 1, strict: true, expectResume: true},
> 		{targetBytes: 11, strict: true, expectResume: false},
> ```

Should we include a `strict: false` variant of each of these two cases?
___
*[pkg/storage/mvcc_history_test.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCLoV89716ydpQJ3rs:-MgCLoV89716ydpQJ3rt:bbie1mr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/storage/mvcc_history_test.go#L59)):*
> ```Go
> // increment [t=<name>] [ts=<int>[,<int>]] [resolve [status=<txnstatus>]] k=<key> [inc=<val>]
> // put       [t=<name>] [ts=<int>[,<int>]] [resolve [status=<txnstatus>]] k=<key> v=<string> [raw]
> // scan      [t=<name>] [ts=<int>[,<int>]] [resolve [status=<txnstatus>]] k=<key> [end=<key>] [inconsistent] [tombstones] [reverse] [failOnMoreRecent] [localUncertaintyLimit=<int>[,<int>]] [max=<max>] [targetbytes=<target>] [stricttargetbytes=<bool>]
> ```

nit: any reason to include a value with this flag? It looks like most boolean args don't do this, so the parsing logic can just look for the inclusion of the key in the args.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCOKGREbWlOJH20a_8:-MgCOKGREbWlOJH20a_9:bl2kxw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/840cc97f8d86f133851fd425a1992e276fd536ef/pkg/storage/pebble_mvcc_scanner.go#L226)):*
> ```Go
> 
> 	var resume *roachpb.Span
> 	if p.maxKeys > 0 && p.results.count == p.maxKeys && p.advanceKey() {
> ```

Does this change imply that this condition could have always been `if p.advanceKey() {`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, and @yuzefovich)

___
*[pkg/roachpb/api.proto, line 2029 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBd8ln8K0cil7oCztv:-MgCjBN1-5Om4HLOGquM:b-167g9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/roachpb/api.proto#L2029)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Right, that might actually come in handy to signal the DistSender that a limit was exceeded too (see comment in PR description). Will look into it.
</blockquote></details>

Quoting the relevant part of the [RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md#avoiding-wasted-work):
```
We're going to make it such that, when a TargetBytes-batch returns no results, it still returns an estimate of the
size of a row. More generally, whenever TargetBytes caused the request to be truncated, the response is going
to include the size of the next row so that the Streamer can use it for subsequent requests.
```
I think the second sentence specifies a nice-to-have (that might be better off in a separate PR).


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> though I think this would also be a good time to think through whether we want StrictTargetBytes = false to allow overshoot even after it has gathered a single key-value.

Yeah, I can have a look at this. It may change a fair bit of the control flow logic, will see how involved it gets.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MguRqJhAoRZrWVbh0Sl:bgq6ag3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1302 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgC5I1f326fgJBemXhb:-MguML1Q9X5bMTVHmQ-_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1302)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, this is a good point. Let's pull out the header into a local variable to avoid the repeat type switch and dynamic dispatch.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1315 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCPzrM9Y4pCZ9GHhdb:-MguOp0P1ap9MLhdpANC:b-qyvucb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvclient/kvcoord/dist_sender.go#L1315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can this condition and the one below both be replaced by `hasResumeSpan`?
</blockquote></details>

I'm not really sure if this logic is sound, as I outlined in the PR description. It assumes that if a response has a resume span it must have exceeded a limit, but I'm not sure if that always holds. I'm going to do another pass over this.
___
*[pkg/kv/kvserver/batcheval/cmd_get_test.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCQwa508gGJotN3lzZ:-MguOPnJ85Vtu6rJBxGW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed8201df18aa390148110ad98568bf17c9b8320d/pkg/kv/kvserver/batcheval/cmd_get_test.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we include a `strict: false` variant of each of these two cases?
</blockquote></details>

Done.
___
*[pkg/roachpb/api.proto, line 2029 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgBd8ln8K0cil7oCztv:-MguPU8a-fx83Ig_aNbV:bk81hgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7076170b8bd38a44a1cabaa768f2590decf622e/pkg/roachpb/api.proto#L2029)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Quoting the relevant part of the [RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md#avoiding-wasted-work):
```
We're going to make it such that, when a TargetBytes-batch returns no results, it still returns an estimate of the
size of a row. More generally, whenever TargetBytes caused the request to be truncated, the response is going
to include the size of the next row so that the Streamer can use it for subsequent requests.
```
I think the second sentence specifies a nice-to-have (that might be better off in a separate PR).
</blockquote></details>

Yeah, but that field may help tighten up the DistSender logic here as well. Doing another pass over this.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MgCOKGREbWlOJH20a_8:-MguR-1tC9ayi3URFK2Q:b7aj0df) ([raw file](https://github.com/cockroachdb/cockroach/blob/840cc97f8d86f133851fd425a1992e276fd536ef/pkg/storage/pebble_mvcc_scanner.go#L226)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this change imply that this condition could have always been `if p.advanceKey() {`?
</blockquote></details>

Yeah, I believe so, unless I've misread anything. The previous logic used `maxKeys>0` as an explicit signal that a limit was exceeded (not necessarily a key limit, it'd update the `maxKeys` field if a byte limit was exceeded), but as far as I can tell that was unnecessary. Worth reviewing this again though.


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
kvserver: avoid exceeding `TargetBytes`

Previously, `Header.TargetBytes` would always overshoot the target, i.e.
a scan would terminate only after the result set hit or exceeded
`TargetBytes`.

This patch changes the behavior to never overshoot, except for when the
first result exceeds `TargetBytes` (to make sure clients can always make
progress). However, this is not backwards-compatible with the 21.2
`DistSender` logic, which relies on responses overshooting `TargetBytes`
when enforcing limits for split batches. It therefore adds a
corresponding `TargetBytesAvoidExcess` version gate that only enables
this for 22.1, and `MVCCScanOptions.TargetBytesAvoidExcess` to vary the
scan behavior based on it.

Release note: None

erikgrinaker committed:
kvserver: add `ResponseHeader.ResumeNextBytes`

This patch adds the field `ResponseHeader.ResumeNextBytes` which may be
populated with the size of the next result in the `ResumeSpan` if
`ResumeReason` is `RESUME_BYTE_LIMIT` (i.e. if `TargetBytes` was
exceeded). This will be needed by an upcoming [parallel get/scan
streaming library][1] to better manage its memory budget.

The field is best-effort, and may be omitted in some cases where the
result size exactly equals `TargetBytes` and we would have to do
additional work to obtain `ResumeNextBytes` (e.g. at the end of a range,
where we would have to send an additional RPC to the next range).

[1]: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210617_index_lookups_memory_limits.md

Release note: None

erikgrinaker reviewed (commented):
I'm going on vacation for the next week, but I've pushed a work-in-progress of all of the requested changes, along with some additional improvements. I'll need to add/update a bunch of tests, tighten up the code, and ensure backwards compatibility with 21.1 (which I think requires us to preserve the current `TargetBytes` behavior of always overshooting until 22.1 unless we add a cluster version for it). Will get to this when I'm back from vacation.

Feel free to ignore or review this at will.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-Mh4l2di8SvT2kMpEVGZ:biqdkzb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 17 of 17 files at r4, 13 of 13 files at r5, 7 of 7 files at r6, 15 of 15 files at r7, 8 of 8 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @itsbilal, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiAH08tNgNCxzdoYZ:-MhEiAH14_AbKsAAHKcJ:b7h0i66) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
> ```Go
> 			// indicates that the request was part of a batch that has already exhausted
> 			// its limit. We still have to actually do the read above though to populate
> 			// ResumeNextBytes.
> ```

Do we want to do this for all Get requests after the batch has exceeded its limit, or only the next one? I'm concerned that this could cause a large batch of `Get` requests to be quadratic in the number of times that they hit the LSM, which seems undesirable.
___
*[pkg/kv/kvserver/batcheval/cmd_resolve_intent_range.go, line 57 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEfDQh5YQ5lF_EFlUA:-MhEfDQh5YQ5lF_EFlUB:bt8zcpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fb9efa5074508df370cfb76c1af5d64592a674b/pkg/kv/kvserver/batcheval/cmd_resolve_intent_range.go#L57)):*
> ```Go
> 		update.EndKey = resumeSpan.Key
> 		reply.ResumeSpan = resumeSpan
> 		// The given MaxSpanRequestKeys really specifies the number of intents
> ```

So you're saying that it would be surprising to callers if they received a `RESUME_INTENT_LIMIT` after providing `MaxSpanRequestKeys`?
___
*[pkg/roachpb/api.proto, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEhIleE-QKIxNoeG6X:-MhEhIleE-QKIxNoeG6Y:b-5x2n0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fb9efa5074508df370cfb76c1af5d64592a674b/pkg/roachpb/api.proto#L68)):*
> ```protobuf
>   RESUME_UNKNOWN = 0;
>   // A key limit was exceeded.
>   RESUME_KEY_LIMIT = 1;
> ```

Did you consider validating proper use of this in one way or another in the mvcc_history_test.go framework? We already print resume spans, so we could start printing the reason alongside them.
___
*[pkg/roachpb/api.proto, line 120 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiXBb7v4bqA7lq-Go:-MhEiXBb7v4bqA7lq-Gp:bbamp7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/roachpb/api.proto#L120)):*
> ```protobuf
>   // When resume_span is populated, this specifies the size of the next
>   // result entry.
>   int64 resume_next_bytes = 9;
> ```

Should we include this in the mvcc_history_test.go framework as well?
___
*[pkg/storage/testdata/mvcc_histories/max_keys, line 26 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEgSH42JGVpZ3DfXZh:-MhEgSH42JGVpZ3DfXZi:b-2dah4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8fc89b4a17d8bb43b2d7acee27dcefa664a6d86/pkg/storage/testdata/mvcc_histories/max_keys#L26)):*
> ```
> ----
> scan: "a" -> /BYTES/val-a @1.000000000,0
> scan: resume span ["c","z")
> ```

Is this change intentional/correct? Aren't we now skipping the key at "aa"? I must be missing something?


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiAH08tNgNCxzdoYZ:-MhsBolvDDCeZJ_v0WDp:b-7glq8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to do this for all Get requests after the batch has exceeded its limit, or only the next one? I'm concerned that this could cause a large batch of `Get` requests to be quadratic in the number of times that they hit the LSM, which seems undesirable.
</blockquote></details>

It's a good point, I'll revisit this logic. We may not be able to avoid it though -- presumably, if `Get` number 3 would exceed a limit, it's still possible that `Get` number 4 wouldn't, and so we would want to keep trying all of the requests (unless we happen to retrieve the _exact_ number of bytes requested).
___
*[pkg/kv/kvserver/batcheval/cmd_resolve_intent_range.go, line 57 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEfDQh5YQ5lF_EFlUA:-MhsCpTh8nxwprgvFYC0:bchd5w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fb9efa5074508df370cfb76c1af5d64592a674b/pkg/kv/kvserver/batcheval/cmd_resolve_intent_range.go#L57)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

So you're saying that it would be surprising to callers if they received a `RESUME_INTENT_LIMIT` after providing `MaxSpanRequestKeys`?
</blockquote></details>

Yes, the idea was to map the reasons to input parameters -- i.e. if you receive a `RESUME_KEY_LIMIT` then you know it's because of `MaxSpanRequestKeys`. However, we're sort of overloading `MaxSpanRequestKeys` here to mean the number of _intents_ we've scanned, not the number of _keys_ we've scanned. I think this is kind of a bug in itself: if the intents are spread out wide then we might have to scan millions of rows before we found the requested number of intents and returned, but I suppose the separated lock table will address that.
___
*[pkg/roachpb/api.proto, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEhIleE-QKIxNoeG6X:-MhsEehiAAC6DyAjTT-V:bi15w02) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fb9efa5074508df370cfb76c1af5d64592a674b/pkg/roachpb/api.proto#L68)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did you consider validating proper use of this in one way or another in the mvcc_history_test.go framework? We already print resume spans, so we could start printing the reason alongside them.
</blockquote></details>

Yes, I was planning to add that.
___
*[pkg/roachpb/api.proto, line 120 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiXBb7v4bqA7lq-Go:-MhsEjmnBMKmVahFehop:b-ougsr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/roachpb/api.proto#L120)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we include this in the mvcc_history_test.go framework as well?
</blockquote></details>

We should.
___
*[pkg/storage/testdata/mvcc_histories/max_keys, line 26 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEgSH42JGVpZ3DfXZh:-MhsEmbV-anYADh7fjMl:b-perll4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8fc89b4a17d8bb43b2d7acee27dcefa664a6d86/pkg/storage/testdata/mvcc_histories/max_keys#L26)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this change intentional/correct? Aren't we now skipping the key at "aa"? I must be missing something?
</blockquote></details>

The key at "aa" is not visible, as it was deleted at ts=2. The old scan logic would return a resume span from the next existing key (regardless of whether that key is visible to the caller or satisfied the scan options), the new logic returns a resume span from the next visible key that satisfies the scan options (if any).

This is necessary to get the byte size of the next entry, and even without that it's generally beneficial since it could allow eliminating another scan request that wouldn't return any values.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
I think this should be ready for another look now. Some tests may still fail in CI, we'll see.

There may still be some logic in the DistSender that needs work, to handle `ResumeNextBytes` in `fillSkippedResponses()` (with tests). `TestScanReverseScanTargetBytes` is also ripe for a rewrite, it's grown pretty convoluted. But I figured I'd get some feedback on the rest of the changes in the meanwhile.

I am a bit worried about the amount of complexity we're adding here, but the functional requirements should be satisfied at least. There are also some internal discussions on Slack about just tearing out all of the `DistSender` logic around limits and returning underfull responses at each range boundary, since clients will generally just keep pulling until the resume span is empty anyway: https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1630007804010500

> ensure backwards compatibility with 21.1 (which I think requires us to preserve the current TargetBytes behavior of always overshooting until 22.1 unless we add a cluster version for it

I added a cluster version to ensure backwards compatibility with 21.1. Once upgraded to 21.2, `TargetBytes` will never be overshot except in the case where only a single result is returned.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-Mi72vhY-19v_N0dcOet:b-igge9a)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiAH08tNgNCxzdoYZ:-Mi71hUTAsImD5rqsx5k:bkyr32x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

It's a good point, I'll revisit this logic. We may not be able to avoid it though -- presumably, if `Get` number 3 would exceed a limit, it's still possible that `Get` number 4 wouldn't, and so we would want to keep trying all of the requests (unless we happen to retrieve the _exact_ number of bytes requested).
</blockquote></details>

On second thought, it makes sense to stop evaluation after the first limit is exceeded, since that's generally where the client will resume from anyway.

To avoid the quadratic behavior, I unfortunately had to introduce another field `AllowEmptyResumeNextBytes`. This is needed for the edge case where a request to one range exactly hits `TargetBytes`, so we need the next range to evaluate with a limit of 0 and populate the resumespan and `ResumeNextBytes` -- another option would be to use a magic value of e.g. `-2` for `TargetBytes`, but this seems clearer.

This also changes the behavior such that `ResumeNextBytes` is only returned for the first request in a batch that exceeds a limit.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
I agree with Yahor that at this point, it's probably best to wait for the branch to be cut. There are some subtle conditions in here that would be better to bake off of the release branch if we don't need the rest of this immediately.

> Reviewed 2 of 17 files at r16, 8 of 20 files at r17, 4 of 15 files at r18, 18 of 19 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MiTRAJK3vx6COxyakZi:bh8ckop)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker, @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/clusterversion/cockroach_versions.go, line 571 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTMCSgDqqyhYOww0NP:-MiTMCSgDqqyhYOww0NQ:b-mkkh47) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/clusterversion/cockroach_versions.go#L571)):*
> ```Go
> 	{
> 		Key:     AvoidExcessTargetBytes,
> 		Version: roachpb.Version{Major: 21, Minor: 1, Internal: 162},
> ```

If this PR isn't going to make it into 21.2, we'll need to update this.
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiAH08tNgNCxzdoYZ:-MiTJvdr77Y-8G4D56NV:b-8iznhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
> another option would be to use a magic value of e.g. -2 for TargetBytes, but this seems clearer

This option doesn't sound bad to me, as long as we pull the `-1` and `-2` sentinel values into well-commented constants. `AllowEmptyResumeNextBytes` is a special-case of `TargetBytes = -1`, right?

But I'm also fine with what you have here. Up to you.
___
*[pkg/roachpb/api.proto, line 66 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTLGl41d59Y00TRM5i:-MiTLGl41d59Y00TRM5j:b1rv95) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/roachpb/api.proto#L66)):*
> ```protobuf
> enum ResumeReason {
>   option (gogoproto.goproto_enum_prefix) = false;
>   // Zero value; no resume, or an unknown reason from a future or past version.
> ```

nit: consider `s/version/cockroachdb version/`, or something to avoid confusion with the 10 other uses of the term "version".
___
*[pkg/storage/mvcc.go, line 2501 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTLiP02ZTnSceeMYIk:-MiTLiP02ZTnSceeMYIl:bybahhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/mvcc.go#L2501)):*
> ```Go
> 	// corresponding ResumeSpan is returned. Negative values return an empty
> 	// result with an appropriate resume span, but will still perform a scan to
> 	// populate ResumeNextBytes.
> ```

"... if AllowEmptyResumeNextBytes is not set to true."

Same below.
___
*[pkg/storage/mvcc.go, line 2521 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTLzTH8Q6yrx7FZKhz:-MiTLzTH8Q6yrx7FZKi-:bjzmepm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/mvcc.go#L2521)):*
> ```Go
> 	//
> 	// TODO(erikgrinaker): This option exists for backwards compatibility with
> 	// 21.1 RPC clients, in 22.1 it should always be enabled.
> ```

If this PR isn't going to make it into 21.2, we'll need to update this comment.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 717 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTP2ea1biJpMNhnHLg:-MiTP2ea1biJpMNhnHLh:b-73is86) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/pebble_mvcc_scanner.go#L717)):*
> ```
> p.targetBytesAllowEmpty
> ```

Should this be `(p.targetBytesAllowEmpty && p.targetBytesAvoidExcess)` like the comment above implies?
___
*[pkg/storage/pebble_mvcc_scanner.go, line 717 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTQrlnEEN_DzZKoov6:-MiTQrlnEEN_DzZKoov7:bfpyitu) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/pebble_mvcc_scanner.go#L717)):*
> ```
> || p.targetBytes < 0
> ```

Do you think it would be cleaner to detect this at some higher level and use it to set `p.targetBytesAllowEmpty` to `true`? That seems to be the implied relationship between these two arguments.
___
*[pkg/storage/testdata/mvcc_histories/max_keys, line 26 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEgSH42JGVpZ3DfXZh:-MiTNzEYDala3rqniEKW:b-vcjhub) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8fc89b4a17d8bb43b2d7acee27dcefa664a6d86/pkg/storage/testdata/mvcc_histories/max_keys#L26)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The key at "aa" is not visible, as it was deleted at ts=2. The old scan logic would return a resume span from the next existing key (regardless of whether that key is visible to the caller or satisfied the scan options), the new logic returns a resume span from the next visible key that satisfies the scan options (if any).

This is necessary to get the byte size of the next entry, and even without that it's generally beneficial since it could allow eliminating another scan request that wouldn't return any values.
</blockquote></details>

Got it, thanks for explaining.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
Because of #70564 I'm going to make some changes to the logic here. Specifically, we'll only populate `ResumeNextBytes` when the resume reason is `RESUME_BYTE_LIMIT`, which also reverts the `MaxSpanRequestKeys` resume span behavior such that it no longer eagerly scans up to the next visible key (and saves us the associated read cost).

Since the changes in that issue will have implications for the DistSender limit handling, I'm leaning towards splitting off the first couple of commits from this PR (improving the resume reasons) and completing #70564 first, but we'll see. That's likely to involve its own set of version gates and such. If you have opinions on that issue, this would be a good time to weigh in.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MkCbtldCsDKRq101Tpm:brkzzxe)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 56 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MhEiAH08tNgNCxzdoYZ:-MkCbNQF0mmKPSVp66tr:busq973) ([raw file](https://github.com/cockroachdb/cockroach/blob/b60eafad310adf977e72e829897d6cca5e376f64/pkg/kv/kvserver/batcheval/cmd_get.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> another option would be to use a magic value of e.g. -2 for TargetBytes, but this seems clearer

This option doesn't sound bad to me, as long as we pull the `-1` and `-2` sentinel values into well-commented constants. `AllowEmptyResumeNextBytes` is a special-case of `TargetBytes = -1`, right?

But I'm also fine with what you have here. Up to you.
</blockquote></details>

That's right, both options are viable, but I do feel like an additional field is clearer. I'm going to make some changes to this logic due to #70564 anyway, will see where we end up.
___
*[pkg/storage/mvcc.go, line 2501 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTLiP02ZTnSceeMYIk:-MkCXg6K2S8H9LJ42MJT:bvm05ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/mvcc.go#L2501)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"... if AllowEmptyResumeNextBytes is not set to true."

Same below.
</blockquote></details>

`AllowEmptyResumeNextBytes` is never propagated down to the MVCC level, its purpose is to skip the MVCC scan/get entirely.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 717 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTP2ea1biJpMNhnHLg:-MkC_EXE0tuIGTSXHvN5:byh6430) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/pebble_mvcc_scanner.go#L717)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> p.targetBytesAllowEmpty
> ```

Should this be `(p.targetBytesAllowEmpty && p.targetBytesAvoidExcess)` like the comment above implies?
</blockquote></details>

I don't believe so? This predicate determines whether to apply a `targetBytes` check at all to the first result, while `targetBytesAvoidExcess` modifies the behavior of that check below. Added a comment to clarify this.
___
*[pkg/storage/pebble_mvcc_scanner.go, line 717 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MiTQrlnEEN_DzZKoov6:-MkCaYeg27amb4iaq_30:b-oc0sgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb47f431a9fe4dcaed998fb830ae820b9d50a9df/pkg/storage/pebble_mvcc_scanner.go#L717)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> || p.targetBytes < 0
> ```

Do you think it would be cleaner to detect this at some higher level and use it to set `p.targetBytesAllowEmpty` to `true`? That seems to be the implied relationship between these two arguments.
</blockquote></details>

I don't think that's true. `targetBytes < 0` will _always_ return an empty result, while `targetBytesAllowEmpty` _may_ return an empty result _if_ the first result exceeds `targetBytes`.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
Ok, I think this should be ready for another (final?) look, pending CI. It looks like the range boundary pagination changes in #70763 will end up being optional, so there's no point blocking this on that. However, I have made some changes to the behavior here such that it will integrate better with #70763 if it goes through. Main changes:

* `ResumeNextBytes` is now only populated if the resume reason is `RESUME_BYTE_LIMIT`. In particular, this avoids having to hit another range if #70763 returns `RESUME_RANGE_BOUNDARY`, which would defeat the whole purpose.

* `MaxSpanRequestKeys` no longer returns resume spans starting from the next visible key, since we don't have to populate `ResumeNextBytes`. Instead, it is reverted to the old behavior of returning a resume span from the next possible key. It could still be beneficial to make this change (i.e. eagerly read to the next visible key), since it could avoid additional RPCs in cases where there are no more visible/matching keys in a range, but that's going to need benchmarking to determine whether it's beneficial so we can consider doing that separately.

* `AllowEmptyResumeNextBytes ` is changed to `TargetBytesRequireNextBytes`, since it now only applies to `TargetBytes` limits. It is still backwards compatible because we only set it during local replica evaluation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlFYwgY8HOrGxw8gBFU:b-n6g6q7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
PS: This bumps the binary version from 21.2 to 22.1. Is there a particular process for this, or is what I'm doing here fine?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlFid875hEGzjnyigZE:b-ezscl0)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

tbg commented:
@celiala would know about process. I'm personally a bit surprised we haven't added the 22.1 start cluster version yet, but maybe that has to do with everything being kind of sideways this cycle due to wanting to cut the 21.2 release only very late on the release branch.

irfansharif commented:
We're waiting for a stable RC: https://github.com/cockroachdb/cockroach/pull/69828

erikgrinaker commented:
Thanks, I'll hold off on merging until that's in then.

erikgrinaker reviewed (commented):
> `AllowEmptyResumeNextBytes`  is changed to `TargetBytesRequireNextBytes`, since it now only applies to `TargetBytes` limits. It is still backwards compatible because we only set it during local replica evaluation.

I realized we're not handling this correctly in the DistSender yet. However, I'm starting to wonder if this is even worth it. The only reason we need this field is for the specific case where a partial response (either for a range, or for one of the responses in a batch) _exactly_ matches `TargetBytes`. We then need to do another read (either via RPC or during local evaluation) to obtain `ResumeNextBytes`. However, is it worth incurring additional cost to obtain `ResumeNextBytes` in this case, or should we say that it's best-effort and can be 0 in some rare cases? Since we're paying both a performance cost and complexity cost for it, I'm inclined to drop it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlGXc91CbXVgt-oitOO:b-xlpv3y)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
LGTM from the high level, thanks!

> The only reason we need this field is for the specific case where a partial response (either for a range, or for one of the responses in a batch) exactly matches TargetBytes. We then need to do another read (either via RPC or during local evaluation) to obtain ResumeNextBytes. However, is it worth incurring additional cost to obtain ResumeNextBytes in this case, or should we say that it's best-effort and can be 0 in some rare cases? Since we're paying both a performance cost and complexity cost for it, I'm inclined to drop it.

I think it makes sense to make it best-effort in this case.

I think the desire to know `ResumeNextBytes` is driven by the performance optimization in the scenario when we issue a parallel batch which cannot exceed `TargetBytes` (because there is another batch - which is the "top of the line" - that might exceed it in order to make progress). If the former couldn't return anything in the response due to `TargetBytes` being too small, at least we want to know what is the smallest budget that we should send the same batch again with. However, if the batch response already contains something that happen to use exactly the `TargetBytes` budget, it's not that important to know the size of the next kv to be read in that range.

> Reviewed 2 of 20 files at r22, 5 of 31 files at r25, 1 of 22 files at r28, 22 of 22 files at r31, 15 of 15 files at r32, 17 of 17 files at r33.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlLCMQ81dt939QNYpAP:bx2n5pn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> I think it makes sense to make it best-effort in this case.
>
> I think the desire to know ResumeNextBytes is driven by the performance optimization in the scenario when we issue a parallel batch which cannot exceed TargetBytes (because there is another batch - which is the "top of the line" - that might exceed it in order to make progress). If the former couldn't return anything in the response due to TargetBytes being too small, at least we want to know what is the smallest budget that we should send the same batch again with. However, if the batch response already contains something that happen to use exactly the TargetBytes budget, it's not that important to know the size of the next kv to be read in that range.

Makes sense, thanks @yuzefovich. Doesn't seem worth it then, so I've removed `TargetBytesRequireNextBytes` -- we can always add it later if we find that we need it. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlPhfP17BzIOktfpcHh:b-slkirn)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 20 files at r22, 2 of 31 files at r25, 2 of 23 files at r34, 23 of 23 files at r37, 14 of 14 files at r38, 16 of 16 files at r39.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-:-MlpgdCbEPNSKsPAdLLF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1348 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-Mlpd6I629n7R0jAWlrF:-Mlpd6I629n7R0jAWlrG:bwz8w7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b0f98e8f5f07ed6b2bf2656a4492d299e8d5d2/pkg/kv/kvclient/kvcoord/dist_sender.go#L1348)):*
> ```Go
> 			// Check whether we've received enough responses to exit query loop.
> 			if mightStopEarly {
> 				var replyResults int64
> ```

nit: `s/replyResults/replyKeys/` to line up with the naming of `NumKeys` and `MaxSpanRequestKeys`. Can be in a separate commit if you'd prefer.
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 72 at r38](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MlpfUPa2eNhb9xWPn_u:-MlpfUPa2eNhb9xWPn_v:b4fiv2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0e890bafa83cbd642baacd949886e87fac2873/pkg/kv/kvserver/batcheval/cmd_get.go#L72)):*
> ```Go
> 		// AvoidExcessTargetBytes cluster version is active.
> 		avoidExcess := cArgs.EvalCtx.ClusterSettings().Version.IsActive(ctx,
> 			clusterversion.AvoidExcessTargetBytes)
> ```

Are we only checking the cluster setting when `h.TargetBytesAllowEmpty` is set to true for `Gets`? Don't we also want to do so for `Scans` as well? But actually, why check this at all? If the client is setting `TargetBytesAllowEmpty`, shouldn't we trust that it has already checked the cluster setting, in which case we should over the flag even if our own (asynchronously updated) view of the cluster setting is different?
___
*[pkg/roachpb/api.proto, line 2141 at r38](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-Mlpf7cV2kAjcMnNihzV:-Mlpf7cV2kAjcMnNihzW:b-vzbxgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0e890bafa83cbd642baacd949886e87fac2873/pkg/roachpb/api.proto#L2141)):*
> ```protobuf
>   // If true, allow returning an empty result when the first result exceeds
>   // target_bytes. Only effective on 22.1 clusters, since 21.2 always overshoots
>   // the limit. Only supported by Get and Scan.
> ```

And reverse scan, right?


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
kvclient: rename `replyResults` to `replyKeys`

Release note: None

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1348 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-Mlpd6I629n7R0jAWlrF:-MlqCnmK-0OwEi5XDqmB:b-1eh3wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/17b0f98e8f5f07ed6b2bf2656a4492d299e8d5d2/pkg/kv/kvclient/kvcoord/dist_sender.go#L1348)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `s/replyResults/replyKeys/` to line up with the naming of `NumKeys` and `MaxSpanRequestKeys`. Can be in a separate commit if you'd prefer.
</blockquote></details>

Sure, done.
___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 72 at r38](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MlpfUPa2eNhb9xWPn_u:-MlqCqoK4tRxV7mJwEY9:b3tmlqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0e890bafa83cbd642baacd949886e87fac2873/pkg/kv/kvserver/batcheval/cmd_get.go#L72)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we only checking the cluster setting when `h.TargetBytesAllowEmpty` is set to true for `Gets`? Don't we also want to do so for `Scans` as well? But actually, why check this at all? If the client is setting `TargetBytesAllowEmpty`, shouldn't we trust that it has already checked the cluster setting, in which case we should over the flag even if our own (asynchronously updated) view of the cluster setting is different?
</blockquote></details>

`Scan` has the same check:

https://github.com/cockroachdb/cockroach/blob/698b8b1c569914e00d3d1c4944f781b8ae616748/pkg/storage/pebble_mvcc_scanner.go#L719-L725

Without the cluster setting, `TargetBytes` will _always_  overshoot, and so it will always include at least 1 result. Therefore `TargetBytesAllowEmpty` is logically meaningless without that setting: sure, it _can_ return an empty result, but it never will since it always overshoots. Perhaps it's academic in the case of `Get`, but it isn't for `Scan` where each setting has a separate effect.

As you point out, this means that clients will see both behaviors (overshot and not, empty and not) during the async migration process. I figured that was probably fine in this case. If it isn't, we can add a new request field to control the overshooting instead, and check the cluster version on the client.
___
*[pkg/roachpb/api.proto, line 2141 at r38](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-Mlpf7cV2kAjcMnNihzV:-MlqHANY7CjtsbM3IgAI:b-mp1u8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0e890bafa83cbd642baacd949886e87fac2873/pkg/roachpb/api.proto#L2141)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

And reverse scan, right?
</blockquote></details>

Yep. Figured that was implied, but made it explicit.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
> Reviewed 1 of 20 files at r40, 3 of 14 files at r41, 16 of 16 files at r42, 1 of 1 files at r43.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68370)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal, @nvanbenschoten, and @yuzefovich)

___
*[pkg/kv/kvserver/batcheval/cmd_get.go, line 72 at r38](https://reviewable.io/reviews/cockroachdb/cockroach/68370#-MlpfUPa2eNhb9xWPn_u:-MltSHMTCEgZgckhhGIL:b8g9k32) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0e890bafa83cbd642baacd949886e87fac2873/pkg/kv/kvserver/batcheval/cmd_get.go#L72)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

`Scan` has the same check:

https://github.com/cockroachdb/cockroach/blob/698b8b1c569914e00d3d1c4944f781b8ae616748/pkg/storage/pebble_mvcc_scanner.go#L719-L725

Without the cluster setting, `TargetBytes` will _always_  overshoot, and so it will always include at least 1 result. Therefore `TargetBytesAllowEmpty` is logically meaningless without that setting: sure, it _can_ return an empty result, but it never will since it always overshoots. Perhaps it's academic in the case of `Get`, but it isn't for `Scan` where each setting has a separate effect.

As you point out, this means that clients will see both behaviors (overshot and not, empty and not) during the async migration process. I figured that was probably fine in this case. If it isn't, we can add a new request field to control the overshooting instead, and check the cluster version on the client.
</blockquote></details>

On second thought, I think you're right here. At the RPC level, `TargetBytesAllowEmpty` should take precedence over the `AvoidExcessTargetBytes` cluster setting -- that is the only setting the client has any control over, and if the client specifies it then it's never going to want the old overshooting behavior. I've changed the RPC commands to override/ignore `AvoidExcessTargetBytes` as appropriate, but kept the MVCC/Pebble-level mechanics logically consistent wrt. these two parameters since callers control both here. This will go away in 22.2 anyway, when we remove the overshooting behavior entirely.


<!-- Sent from Reviewable.io -->

irfansharif commented:
> Thanks, I'll hold off on merging until that's in then.

PS: #69828 just landed, which IIUC unblocks you here.

erikgrinaker commented:
>PS: #69828 just landed, which IIUC unblocks you here.

Thanks for the heads-up!

erikgrinaker commented:
bors r=nvanbenschoten,yuzefovich

TFTRs, and apologies for the slow delivery!
